homeassistant/components/kdeconnect/__init__.py

```diff
@@ -0,0 +1,107 @@
+"""The KDEConnect integration."""
+from __future__ import annotations
+
+from abc import ABC, abstractmethod
+import logging
+from typing import cast
+
+from pykdeconnect.client import KdeConnectClient
+from pykdeconnect.devices import KdeConnectDevice
+
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import CONF_DEVICE_ID, Platform
+from homeassistant.core import HomeAssistant, State
+from homeassistant.helpers.entity import DeviceInfo
+from homeassistant.helpers.restore_state import RestoreEntity
+from homeassistant.helpers.typing import ConfigType
+
+from .const import DATA_KEY_CLIENT, DATA_KEY_DEVICES, DATA_KEY_STORAGE, DOMAIN
+from .helpers import ensure_running
+from .storage import HomeAssistantStorage
+
+_LOGGER = logging.getLogger(__name__)
+
+
+PLATFORMS: list[Platform] = [
+    # Platform.BUTTON,
+    Platform.BINARY_SENSOR,
+    Platform.SENSOR,
+    # Platform.DEVICE_TRACKER,
+    # Platform.MEDIA_PLAYER,
+]
+
+
+async def async_setup(hass: HomeAssistant, config: ConfigType) -> bool:
+    """Set up the KDE Connect integration."""
+    await ensure_running(hass)
+
+    if DOMAIN in config:
+        _LOGGER.warning("%s can't be set up using the configuration file", DOMAIN)
+
+    return True
+
+
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Set up KDE Connect from a config entry."""
+    client = cast(KdeConnectClient, hass.data[DOMAIN][DATA_KEY_CLIENT])
+    storage = cast(HomeAssistantStorage, hass.data[DOMAIN][DATA_KEY_STORAGE])
+
+    storage.add_device(entry)
+
+    device_id = entry.data[CONF_DEVICE_ID]
+    device = client.get_device(device_id)
+    assert device is not None
+
+    hass.data[DOMAIN][DATA_KEY_DEVICES][entry.entry_id] = device
```

Comment:
```suggestion
    hass.data.setdefault(DOMAIN, {})[entry.entry_id][DATA_KEY_DEVICES] = device
```

Reply:
`entry.entry_id` should go before `DATA_KEY_DEVICES`. If only `device` needs to be stored, then we don't need any key at all.

Reply:
I also need to store data that's shared between all entries (DATA_KEY_INSTANCE and DATA_KEY_STORAGE). Those two are directly stored in the DOMAIN dict, so I moved the entries into their own dict (DATA_KEY_DEVICES).
It doesn't seem right to store all devices in the same dict as a single client and storage instance.

I can still use setdefault in ensure_running though.